### PR TITLE
WIP: Oscar 2.0

### PR DESCRIPTION
**Pull request by maerteijn**
Checklist  for Oscar 2.0 compatibility.
- [x] Fix nasty travis ci segmentation fault when running the tests (use newer sqlite3)
- [x] Fix issue with oscar main template path
- [x] Test with django 2.2 and remove 1.11 compatibility
- [x] Update middleware classes
- [x] Remove six (no python 2 support needed anymore)
- [x] Replace urls with paths
- [x] Remove PutIsPatchMixin
- [x] Fix sphinx docs building with newer versions of django
- [x] Remove the demosite  or rename to override-example
- [x] Update documentation with supported dependencies
- [x] Update documentation with urls/paths and how to override url's
- [x]  Update README
- [x] ~Move the url config from urls.py to the `OscarConfig` app~ Remove unused `OscarConfig` app and use django's build in `AppConfig`
- [x] ~Test Admin AP specific unique code with new Django 2.2 constraints~ Add issue that some work needs to be done to be compatible with Django 2.2+ `UniqueConstrains`.

### CODE REVIEW

**Comment by Andruten** [comment]
Hi! Do you accept contributions? i could help you with this or #132

**Comment by specialunderwear** [comment]
Yes ofcourse we accept contributions, work on #132 would be much appreciated

**Comment by maerteijn** [comment]
BTW, Do not merge yet.

**Comment by specialunderwear** [comment]
In the tests, "put" has been changed to "patch" in some places. I was wondering why that is? Where some tests not passing with "put"?

**Comment by maerteijn** [comment]
> In the tests, "put" has been changed to "patch" in some places. I was wondering why that is? Where some tests not passing with "put"?

This is because of the removal of `PutIsPatchMixin`. So you need patch for partial updated now.

File: README.rst [hunk 1]
```diff
@@ -33,26 +33,11 @@ steps:
...
-
-      from django.urls import path
-
-      from oscarapi.app import application as api
+      from django.urls import include
 
       urlpatterns = (
           # all the things you already have
-          path('oscarapi/', api.urls),
+          path("api/", include("oscarapi.urls")),
```
Comment (specialunderwear) [inline README.rst]: Is it now nolonger possible to disable some functionality that is not needed?

Comment (specialunderwear) [reply README.rst]: i used to do that by filtering urls in the app config.

Comment (maerteijn) [reply README.rst]: I see now that `OscarConfigMixin` also has some other url related methods as well, so I should use the app for returning the url's indeed.

**Comment by maerteijn** [comment]
@specialunderwear 
I did some research on the newer django 2.2 constraints, but these are not automatically mapped to `_meta.unique_together`.

Which would mean that:
```python
def construct_id_filter(model, data, prefix=None):
    _filter = models.Q()

    for unique_together in model._meta.unique_together:
        _filter |= models.Q
         ...
```
now used in oscarapi will not detect these newly `UniqueConstraints` options. As `unique_together` could be marked [deprecated](https://docs.djangoproject.com/en/2.2/ref/models/options/#unique-together) in the future: What should we do with this?

**Comment by specialunderwear** [comment]
We only have to do something when oscar starts using that feature. We are tied to the models in oscar. Do we have a nighly build? It will start breaking as soon as they implement that.

**Comment by maerteijn** [comment]
(And when people override the models and use these `UniqueConstraints` but this is something we can document)

**Comment by specialunderwear** [comment]
It is most likely that they will add constraints instead of remove existing constraints, so it is very unlikely that something will break. However you can just add code to construct_id_filter that handles the new constraints if you want, I don't think that it will be hard. You don't have to change anything just add stuff.

**Comment by maerteijn** [comment]
Well not really as these newly constraints can be quite complicated in a way like "only unique when values are in these range".

I don't  have a use-case for this feature right now so I won't  spend my time on it. I'll add a section to the documentation and an issue to the tracker that this needs to be done sometime.

File: README.rst (hunk 1 above)
Comment (maerteijn) [reply README.rst]: For now you need to create your own `urlpatterns` if you want this changed. But this was the case  also in the old situation.

Agreed to make it simpler like this?

**Comment by maerteijn** [comment]
I consider this pull request now ready for merge. @specialunderwear Could you review once more or merge it when you agree?

**Comment by specialunderwear** [comment]
I think we need to run black over all the source files and be done with it.
Let's do it in this release.

**Comment by specialunderwear** [comment]
black is beautiful

**Comment by prince-tanapong** [comment]
Have a plane to merge this?

**Comment by maerteijn** [comment]
I’ll apply black in a seperate pull request, so ready for merge special underwear?

**Comment by specialunderwear** [comment]
y

**Comment by prince-tanapong** [comment]
When you guys have a plan to release this?

**Comment by maerteijn** [comment]
Soonish

**Pull request merged**